### PR TITLE
ICU-21840 Fix formatting Aliases

### PR DESCRIPTION
**Pull request by younies**
##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21840
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/test/intltest/measfmttest.cpp [hunk 1]
```diff
@@ -5066,6 +5068,30 @@ void MeasureFormatTest::TestInternalMeasureUnitImpl() {
...
+    auto milligram_per_deciliter_new3 = MeasureUnit::createMilligramOfglucosePerDeciliter(status);
+
+    assertSuccess("test", status);
+    assertEquals("test", milligram_per_deciliter_new.getIdentifier(),
+                 "milligram-ofglucose-per-deciliter");
+    assertEquals("test", milligram_per_deciliter_new2.getIdentifier(),
+                 "milligram-ofglucose-per-deciliter");
+    assertEquals("test", milligram_per_deciliter_new3->getIdentifier(),
+                 "milligram-ofglucose-per-deciliter");
+}
```
Comment (pedberg-icu) [inline icu4c/source/test/intltest/measfmttest.cpp:5095]: These tests do not test the problem described in the bug. That problem was about errors when *formatting* using e.g. milligram_per_deciliter or milligram_per_deciliter2 or milligram_per_deciliter3 as created above. It was not about problems in creating or equating the MeasureUnits themselves.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 2]
```diff
@@ -4424,6 +4424,21 @@ public void TestInternalMeasureUnitImpl() {
...
+
+        MeasureUnit milligramPerDeciliterNew = MeasureUnit.forIdentifier("milligram-ofglucose-per-deciliter");
+        MeasureUnit milligramPerDeciliterNew2 = MeasureUnit.MILLIGRAM_OFGLUCOSE_PER_DECILITER;
+        assertEquals("test", milligramPerDeciliterNew.getIdentifier(), "milligram-ofglucose-per-deciliter");
+        assertTrue("test", milligramPerDeciliterNew.equals(milligramPerDeciliterNew2));
+
+        assertFalse("test", milligramPerDeciliter.equals(milligramPerDeciliterNew));
+    }
+
     private void verifyCompoundUnit(
```
Comment (pedberg-icu) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:4442]: Same comment as above

**Review by pedberg-icu — CHANGES_REQUESTED** [review]

File: icu4c/source/test/intltest/measfmttest.cpp (hunk 1 above)
Comment (younies) [reply icu4c/source/test/intltest/measfmttest.cpp]: Acknowledged.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java (hunk 2 above)
Comment (younies) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java]: Acknowledged.

File: icu4c/source/i18n/number_longnames.cpp [hunk 3]
```diff
@@ -431,13 +431,20 @@ void getMeasureData(const Locale &locale,
     subKey.append(unit.getType(), status);
     subKey.append("/", status);
 
+    // TODO(ICU-21938): remove after fixing the data in CLDR-15421
+    const char *unitSubType = unit.getSubtype();
```
Comment (pedberg-icu) [inline icu4c/source/i18n/number_longnames.cpp:435]: CLDR-15241 is not going to be fixed, CLDR is behaving as intended, see comments in that ticket. Any fix here needs to be a real fix, not a hack.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java [hunk 4]
```diff
@@ -346,12 +346,19 @@ static void getMeasureData(
         subKey.append(unit.getType());
         subKey.append("/");
 
+        //TODO(ICU-21938): remove after fixing the data in CLDR-15421
```
Comment (pedberg-icu) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java:349]: CLDR-15241 is not going to be fixed, CLDR is behaving as intended, see comments in that ticket. Any fix here needs to be a real fix, not a hack.

**Review by pedberg-icu — CHANGES_REQUESTED** [review]

File: icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java (hunk 4 above)
Comment (macchiati) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java:349]: Yes, the problem is that ICU is not using the alias mapping from CLDR (see https://github.com/unicode-org/cldr/blob/main/common/supplemental/units.xml#L485) in order to map deprecated unit identifiers to the new ones.

Comment (younies) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java:349]: Okay, no we need to map the aliases before we look to it. However, those data is not in added yet and I need to added. see

Comment (pedberg-icu) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java:349]: I will try to fix that mapping of alias data when I do the CLDR  beta1 to ICU integration starting today.

Comment (younies) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java:349]: I already wrote the rules to add the aliases, I am still figuring out how to run the tool.
https://github.com/unicode-org/icu/pull/2026

I will try to finish it by tomorrow and you can review it if you do not mind.

File: icu4c/source/i18n/number_longnames.cpp [hunk 5]
```diff
@@ -431,13 +431,39 @@ void getMeasureData(const Locale &locale,
...
+    if (!U_FAILURE(aliasStatus)) {
+        int32_t len;
+        UBool forceCopy = true;
+        unitSubTypeCharString.clear();
+        ures_getUTF8String(aliasFillIn.getAlias(), unitSubTypeCharString.data(), &len, forceCopy,
+                           &aliasStatus);
+    }
+
+    const char *unitSubType = unitSubTypeCharString.data();
+
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:458]: @pedberg-icu , I think there is a better way in C++ to get the data for aliases that what I am doing here. I am not familiar with getting the data in C++. I want to ask you if you think there is a better way.

File: icu4c/source/i18n/number_longnames.cpp [hunk 6]
```diff
@@ -431,13 +431,39 @@ void getMeasureData(const Locale &locale,
...
     // Map duration-year-person, duration-week-person, etc. to duration-year, duration-week, ...
     // TODO(ICU-20400): Get duration-*-person data properly with aliases.
-    int32_t subtypeLen = static_cast<int32_t>(uprv_strlen(unit.getSubtype()));
-    if (subtypeLen > 7 && uprv_strcmp(unit.getSubtype() + subtypeLen - 7, "-person") == 0) {
-        subKey.append({unit.getSubtype(), subtypeLen - 7}, status);
+    // TODO(ICU-21938): use units.getSubType() function instead of unitSubType.
```
Comment (younies) [inline icu4c/source/i18n/number_longnames.cpp:461]: @younies , remove!

**Review by pedberg-icu — DISMISSED** [review]

File: icu4c/source/i18n/number_longnames.cpp (hunk 5 above)
Comment (pedberg-icu) [reply icu4c/source/i18n/number_longnames.cpp:458]: > I will think about that a bit. In meetings now for 3 hours. Adding @richgillam as reviewer, he may have good ideas about this

Looking at this again, I do not know of a better way to get the alias data, I would go ahead with this approach.

**Comment by pedberg-icu** [comment]
Hmm, error in the clang/asan run:
```
formatUnitsAliases {
[9887](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9887)
            FAIL: test unit aliases; got ""; expected "2 milligrams per deciliter"
[9888](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9888)
            FAIL: test unit aliases; got ""; expected "2 liters per 100 kilometers"
[9889](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9889)
            FAIL: test unit aliases; got ""; expected "2 parts per million"
[9890](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9890)
            FAIL: test unit aliases; got ""; expected "2 millimeters of mercury"
[9891](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9891)
            FAIL: test unit aliases; got ""; expected "2 milligrams per deciliter"
[9892](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9892)
            FAIL: test unit aliases; got ""; expected "2 millimeters of mercury"
[9893](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9893)
            FAIL: test unit aliases; got ""; expected "2 liters per 100 kilometers"
[9894](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9894)
            FAIL: test unit aliases; got ""; expected "2 parts per million"
[9895](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9895)
            formatUnitsAliases destructor: expected success but got error: U_INTERNAL_PROGRAM_ERROR
[9896](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9896)
         
[9897](https://github.com/unicode-org/icu/runs/5539798225?check_suite_focus=true#step:3:9897)
            } ERRORS (9) in formatUnitsAliases
```

File: icu4c/source/i18n/number_longnames.cpp (hunk 5 above)
Comment (younies) [reply icu4c/source/i18n/number_longnames.cpp:458]: OKay, I will fix the clang asan and then send it back to you

Comment (younies) [reply icu4c/source/i18n/number_longnames.cpp:458]: thanks :)

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java [hunk 7]
```diff
@@ -5857,6 +5857,45 @@ public void validRanges() throws NoSuchMethodException, IllegalAccessException {
...
+                new TestCase(MeasureUnit.MILLIMETER_OF_MERCURY, "2 millimeters of mercury"),
+
+                // Replacements
+                new TestCase(MeasureUnit.MILLIGRAM_OFGLUCOSE_PER_DECILITER, "2 milligrams per deciliter"),
+                new TestCase(MeasureUnit.forIdentifier("millimeter-ofhg"), "2 millimeters of mercury"),
+                new TestCase(MeasureUnit.forIdentifier("liter-per-100-kilometer"), "2 liters per 100 kilometers"),
+                new TestCase(MeasureUnit.forIdentifier("permillion"), "2 parts per million"),
+        };
+
+        for (int i = 0; i < testCases.length; i++) {
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java:5887]: as I recall, this for loop is not supported by Java 7.0. it is supported starting from Java 8.0

[the remark this answers is not in the document]
Comment (younies) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/number/NumberFormatterApiTest.java]: Done, I think it was related to using other collections than arrays.

**Review by pedberg-icu — APPROVED** [review]

**Pull request merged**